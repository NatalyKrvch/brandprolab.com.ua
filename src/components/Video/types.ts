export interface VideoProps {
  videoUrl: string;
  thumbnailSrc: string;
  className?: string;
}
