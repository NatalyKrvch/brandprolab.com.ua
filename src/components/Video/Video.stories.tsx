import type { Meta, StoryObj } from '@storybook/react';

import Video from './Video';

const meta: Meta<typeof Video> = {
  title: 'Components/Video',
  component: Video,
  argTypes: {
    videoUrl: { control: 'text' },
    thumbnailSrc: { control: 'text' },
    className: { control: 'text' },
  },
  args: {
    videoUrl: 'https://www.youtube.com/watch?v=dQw4w9WgXcQ',
    thumbnailSrc: 'https://img.youtube.com/vi/dQw4w9WgXcQ/maxresdefault.jpg',
    className: '',
  },
};

export default meta;

type Story = StoryObj<typeof Video>;

export const Default: Story = {
  args: {
    videoUrl: 'https://www.youtube.com/watch?v=dQw4w9WgXcQ',
    thumbnailSrc: 'https://img.youtube.com/vi/dQw4w9WgXcQ/maxresdefault.jpg',
    className: '',
  },
};
