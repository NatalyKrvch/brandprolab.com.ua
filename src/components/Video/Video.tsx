'use client';

import clsx from 'clsx';
import Image from 'next/image';
import { twMerge } from 'tailwind-merge';

import {
  VIDEO_IFRAME_TEST_ID,
  VIDEO_PLAY_BUTTON_TEST_ID,
  VIDEO_TEST_ID,
  VIDEO_THUMBNAIL_TEST_ID,
} from '@/lib/testIDs';
import { ROUNDED_CLASSES } from '@/styles/constants';

import { useVideoPlayer } from './hooks/useVideoPlayer';
import type { VideoProps } from './types';

const Video = ({ videoUrl, thumbnailSrc, className }: VideoProps) => {
  const { isPlaying, handlePlay, embedUrl } = useVideoPlayer(videoUrl);

  return (
    <div
      className={twMerge(
        `relative h-210 w-280 overflow-hidden tablet:h-300 tablet:w-400 desktop:h-360 desktop:w-480 ${ROUNDED_CLASSES}`,
        className,
      )}
      data-testid={VIDEO_TEST_ID}
    >
      <div
        className={clsx(
          'absolute inset-0 transition-opacity duration-500',
          isPlaying ? 'pointer-events-none opacity-0' : 'opacity-100',
        )}
      >
        <div className="absolute inset-0 bg-teal-fogOpacity opacity-50" />

        <Image
          src={thumbnailSrc}
          alt=""
          aria-hidden="true"
          loading="lazy"
          width={280}
          height={210}
          className="h-full w-full object-cover"
          data-testid={VIDEO_THUMBNAIL_TEST_ID}
        />

        <div className="absolute inset-0 flex items-center justify-center">
          <button
            className="rounded-full shadow-lg transition hover:scale-105"
            onClick={handlePlay}
            aria-label="Відтворити відео"
            data-testid={VIDEO_PLAY_BUTTON_TEST_ID}
          >
            <Image
              aria-hidden="true"
              src="/icons/nav/play.svg"
              alt=""
              width={74}
              height={74}
            />
          </button>
        </div>
      </div>

      {isPlaying && (
        <iframe
          className="h-full w-full"
          src={embedUrl}
          title="Відео"
          allow="autoplay; encrypted-media"
          allowFullScreen
          data-testid={VIDEO_IFRAME_TEST_ID}
        />
      )}
    </div>
  );
};

export default Video;
