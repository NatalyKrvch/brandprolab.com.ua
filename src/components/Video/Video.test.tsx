import '@testing-library/jest-dom';

import { fireEvent, render, screen } from '@testing-library/react';

import {
  VIDEO_IFRAME_TEST_ID,
  VIDEO_PLAY_BUTTON_TEST_ID,
  VIDEO_TEST_ID,
  VIDEO_THUMBNAIL_TEST_ID,
} from '@/lib/testIDs';

import Video from './Video';

jest.mock('next/image', () => {
  const MockImage = ({ src, alt, ...props }: { src: string; alt: string }) => (
    <img src={src} alt={alt} {...props} />
  );
  MockImage.displayName = 'MockImage';
  return MockImage;
});

describe('Video component', () => {
  const defaultProps = {
    videoUrl: 'https://www.youtube.com/watch?v=testvideo12345',
    thumbnailSrc: '/test-thumbnail.jpg',
  };

  it('renders the Video component', () => {
    const { asFragment } = render(<Video {...defaultProps} />);
    const videoElement = screen.getByTestId(VIDEO_TEST_ID);

    expect(videoElement).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the thumbnail correctly', () => {
    const { asFragment } = render(<Video {...defaultProps} />);
    const thumbnailElement = screen.getByTestId(VIDEO_THUMBNAIL_TEST_ID);

    expect(thumbnailElement).toBeInTheDocument();
    expect(thumbnailElement).toHaveAttribute('src', '/test-thumbnail.jpg');
    expect(thumbnailElement).toHaveAttribute('alt', 'Video Thumbnail');
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the play button', () => {
    const { asFragment } = render(<Video {...defaultProps} />);
    const playButton = screen.getByTestId(VIDEO_PLAY_BUTTON_TEST_ID);

    expect(playButton).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('replaces thumbnail with iframe when play button is clicked', () => {
    const videoUrl = 'https://www.youtube.com/watch?v=testvideo12345';
    const { asFragment } = render(
      <Video videoUrl={videoUrl} thumbnailSrc="/test-thumbnail.jpg" />,
    );
    const playButton = screen.getByTestId(VIDEO_PLAY_BUTTON_TEST_ID);

    fireEvent.click(playButton);

    const iframeElement = screen.getByTestId(VIDEO_IFRAME_TEST_ID);
    expect(iframeElement).toBeInTheDocument();
    expect(iframeElement).toHaveAttribute(
      'src',
      'https://www.youtube.com/embed/testvideo12345?autoplay=1&rel=0',
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('does not render iframe initially', () => {
    const { asFragment } = render(<Video {...defaultProps} />);
    expect(screen.queryByTestId(VIDEO_IFRAME_TEST_ID)).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
