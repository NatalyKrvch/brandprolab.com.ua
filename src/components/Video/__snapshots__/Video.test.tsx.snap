// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video component does not render iframe initially 1`] = `
<DocumentFragment>
  <div
    class="relative h-210 w-280 overflow-hidden tablet:h-300 tablet:w-400 desktop:h-360 desktop:w-480 rounded-24 tablet:rounded-32 desktop:rounded-40"
    data-testid="video-test-id"
  >
    <div
      class="absolute inset-0 transition-opacity duration-500 opacity-100"
    >
      <div
        class="absolute inset-0 bg-teal-fogOpacity opacity-50"
      />
      <img
        alt=""
        aria-hidden="true"
        class="h-full w-full object-cover"
        data-testid="video-thumbnail-test-id"
        height="210"
        loading="lazy"
        src="/test-thumbnail.jpg"
        width="280"
      />
      <div
        class="absolute inset-0 flex items-center justify-center"
      >
        <button
          aria-label="Відтворити відео"
          class="rounded-full shadow-lg transition hover:scale-105"
          data-testid="video-play-button-test-id"
        >
          <img
            alt=""
            aria-hidden="true"
            height="74"
            src="/icons/nav/play.svg"
            width="74"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Video component renders the Video component 1`] = `
<DocumentFragment>
  <div
    class="relative h-210 w-280 overflow-hidden tablet:h-300 tablet:w-400 desktop:h-360 desktop:w-480 rounded-24 tablet:rounded-32 desktop:rounded-40"
    data-testid="video-test-id"
  >
    <div
      class="absolute inset-0 transition-opacity duration-500 opacity-100"
    >
      <div
        class="absolute inset-0 bg-teal-fogOpacity opacity-50"
      />
      <img
        alt=""
        aria-hidden="true"
        class="h-full w-full object-cover"
        data-testid="video-thumbnail-test-id"
        height="210"
        loading="lazy"
        src="/test-thumbnail.jpg"
        width="280"
      />
      <div
        class="absolute inset-0 flex items-center justify-center"
      >
        <button
          aria-label="Відтворити відео"
          class="rounded-full shadow-lg transition hover:scale-105"
          data-testid="video-play-button-test-id"
        >
          <img
            alt=""
            aria-hidden="true"
            height="74"
            src="/icons/nav/play.svg"
            width="74"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Video component renders the play button 1`] = `
<DocumentFragment>
  <div
    class="relative h-210 w-280 overflow-hidden tablet:h-300 tablet:w-400 desktop:h-360 desktop:w-480 rounded-24 tablet:rounded-32 desktop:rounded-40"
    data-testid="video-test-id"
  >
    <div
      class="absolute inset-0 transition-opacity duration-500 opacity-100"
    >
      <div
        class="absolute inset-0 bg-teal-fogOpacity opacity-50"
      />
      <img
        alt=""
        aria-hidden="true"
        class="h-full w-full object-cover"
        data-testid="video-thumbnail-test-id"
        height="210"
        loading="lazy"
        src="/test-thumbnail.jpg"
        width="280"
      />
      <div
        class="absolute inset-0 flex items-center justify-center"
      >
        <button
          aria-label="Відтворити відео"
          class="rounded-full shadow-lg transition hover:scale-105"
          data-testid="video-play-button-test-id"
        >
          <img
            alt=""
            aria-hidden="true"
            height="74"
            src="/icons/nav/play.svg"
            width="74"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Video component renders the thumbnail correctly 1`] = `
<DocumentFragment>
  <div
    class="relative h-210 w-280 overflow-hidden tablet:h-300 tablet:w-400 desktop:h-360 desktop:w-480 rounded-24 tablet:rounded-32 desktop:rounded-40"
    data-testid="video-test-id"
  >
    <div
      class="absolute inset-0 transition-opacity duration-500 opacity-100"
    >
      <div
        class="absolute inset-0 bg-teal-fogOpacity opacity-50"
      />
      <img
        alt=""
        aria-hidden="true"
        class="h-full w-full object-cover"
        data-testid="video-thumbnail-test-id"
        height="210"
        loading="lazy"
        src="/test-thumbnail.jpg"
        width="280"
      />
      <div
        class="absolute inset-0 flex items-center justify-center"
      >
        <button
          aria-label="Відтворити відео"
          class="rounded-full shadow-lg transition hover:scale-105"
          data-testid="video-play-button-test-id"
        >
          <img
            alt=""
            aria-hidden="true"
            height="74"
            src="/icons/nav/play.svg"
            width="74"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Video component replaces thumbnail with iframe when play button is clicked 1`] = `
<DocumentFragment>
  <div
    class="relative h-210 w-280 overflow-hidden tablet:h-300 tablet:w-400 desktop:h-360 desktop:w-480 rounded-24 tablet:rounded-32 desktop:rounded-40"
    data-testid="video-test-id"
  >
    <div
      class="absolute inset-0 transition-opacity duration-500 pointer-events-none opacity-0"
    >
      <div
        class="absolute inset-0 bg-teal-fogOpacity opacity-50"
      />
      <img
        alt=""
        aria-hidden="true"
        class="h-full w-full object-cover"
        data-testid="video-thumbnail-test-id"
        height="210"
        loading="lazy"
        src="/test-thumbnail.jpg"
        width="280"
      />
      <div
        class="absolute inset-0 flex items-center justify-center"
      >
        <button
          aria-label="Відтворити відео"
          class="rounded-full shadow-lg transition hover:scale-105"
          data-testid="video-play-button-test-id"
        >
          <img
            alt=""
            aria-hidden="true"
            height="74"
            src="/icons/nav/play.svg"
            width="74"
          />
        </button>
      </div>
    </div>
    <iframe
      allow="autoplay; encrypted-media"
      allowfullscreen=""
      class="h-full w-full"
      data-testid="video-iframe-test-id"
      src="https://www.youtube.com/embed/testvideo12345?autoplay=1&rel=0"
      title="Відео"
    />
  </div>
</DocumentFragment>
`;
